// @convex-dev/eslint-plugin/dist/commonjs/lib/no-import-use-node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.noImportUseNode = void 0;
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const util_js_1 = require("../util.js");
exports.noImportUseNode = (0, util_js_1.createRule)({
    name: "no-import-use-node",
    meta: {
        type: "suggestion",
        docs: {
            description: 'Only "use node" modules can import other "use node" modules',
        },
        messages: {
            "wrong runtime import": 'This file uses the Convex JavaScript runtime but it imports a "use node" module.',
        },
        schema: [],
    },
    defaultOptions: [],
    create: (context) => {
        const filename = context.getFilename();
        const entry = (0, util_js_1.isEntryPoint)(context.getFilename());
        if (!entry)
            return {};
        const currentDir = path_1.default.dirname(filename);
        return {
            ImportDeclaration(node) {
                if (typeof node.source.value !== "string")
                    return {};
                const relative = node.source.value;
                if (!relative.startsWith("."))
                    return {};
                const abs = path_1.default.resolve(currentDir, relative);
                // TODO this is a heuristic, find out about convex.json
                if (!abs.includes("convex/"))
                    return {};
                const sourceFile = resolveFile(abs);
                if (!sourceFile)
                    return;
                let source;
                try {
                    source = fs_1.default.readFileSync(sourceFile, { encoding: "utf-8" });
                }
                catch {
                    return;
                }
                if (source && source.slice(0, 100).includes("use node")) {
                    context.report({
                        messageId: "wrong runtime import",
                        node: node,
                    });
                }
            },
        };
    },
});
// Implement basic module resolution for relative paths only.
// This doesn't work with path aliases and so many other cases;
// it's a proof of concept that might be helpful to folks debugging.
function resolveFile(filePath) {
    const extensions = [".ts", ".tsx", ".js", ".jsx", ""];
    for (const ext of extensions) {
        const fullPath = `${filePath}${ext}`;
        if (fs_1.default.existsSync(fullPath) && fs_1.default.statSync(fullPath).isFile()) {
            return fullPath;
        }
    }
    // Check for directory import
    if (fs_1.default.existsSync(filePath) && fs_1.default.statSync(filePath).isDirectory()) {
        for (const ext of extensions) {
            const indexPath = path_1.default.join(filePath, `index${ext}`);
            if (fs_1.default.existsSync(indexPath) && fs_1.default.statSync(indexPath).isFile()) {
                return indexPath;
            }
        }
    }
    return null;
}
