// @convex-dev/eslint-plugin/dist/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const no_import_use_node_js_1 = require("./lib/no-import-use-node.js");
const no_old_registered_function_syntax_js_1 = require("./lib/no-old-registered-function-syntax.js");
const require_args_validator_js_1 = require("./lib/require-args-validator.js");
const explicitTableIds_js_1 = require("./lib/explicitTableIds.js");
const version_js_1 = require("./version.js");
const rules = {
    "no-old-registered-function-syntax": no_old_registered_function_syntax_js_1.noOldRegisteredFunctionSyntax,
    "require-args-validator": require_args_validator_js_1.requireArgsValidator,
    "import-wrong-runtime": no_import_use_node_js_1.noImportUseNode,
    "explicit-table-ids": explicitTableIds_js_1.explicitTableIds,
};
const recommendedRules = {
    // This rule is a good idea but bothersome to convert projects to later:
    // it's possible to safely import specific exports from a "use node"
    // file if all Node.js-specific imports are side-effect free.
    "@convex-dev/import-wrong-runtime": "off",
    "@convex-dev/no-old-registered-function-syntax": "error",
    "@convex-dev/require-args-validator": "error",
    "@convex-dev/explicit-table-ids": "error",
};
// Bun is hard to feature detect for ESM vs CJS, so only support ESLint 9 with Bun (contributions welcome)
// @ts-expect-error Bun types are not installed
const isBun = typeof Bun !== "undefined";
// Detect ESM to guess at which ESLint version we're using.
const isESM = typeof require === "undefined" || isBun;
// Base plugin structure, common across ESLint 8 and 9
const plugin = {
    // loose types so this can work with ESlint 8 and 9
    configs: {},
    meta: {
        name: "@convex-dev/eslint-plugin",
        version: version_js_1.version,
    },
    rules,
    processors: {},
};
// ESLint 9 format (ESM)
if (isESM) {
    Object.assign(plugin.configs, {
        recommended: [
            {
                files: ["**/convex/**/*.ts"],
                plugins: {
                    // We could call it "convex" instead, but in ESLint 8 rules can't be renamed like this.
                    // For consistency use @convex-dev/rule-name in ESLint 8 and 9.
                    "@convex-dev": plugin,
                },
                rules: recommendedRules,
            },
        ],
    });
}
// ESLint 8 format (CommonJS)
else {
    plugin.configs = {
        recommended: {
            // Naming for plugins in namespaced packages is special: it removes the "eslint-plugin" part
            // "plugins": [
            //   "jquery", // means eslint-plugin-jquery
            //   "@jquery/jquery", // means @jquery/eslint-plugin-jquery
            //   "@foobar" // means @foobar/eslint-plugin
            // ]
            // Naming for configs in namespaced packages is also special, but this isn't a config.
            plugins: ["@convex-dev"],
            // Apply no rules globally
            rules: {},
            overrides: [
                {
                    // Apply recommended rules in the convex directory
                    files: ["**/convex/**/*.ts"],
                    rules: recommendedRules,
                },
            ],
        },
    };
    // In CommonJS, we need to directly assign to module.exports
    module.exports = plugin;
}
// For ESM (ESLint 9)
exports.default = plugin;
